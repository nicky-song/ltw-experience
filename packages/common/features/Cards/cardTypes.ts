import { Descendant } from 'slate';
import { CardType } from '../../constants';

export type GetCardsParams = {
  learningItemId: string | undefined;
  cardId?: string;
  resetSelectedCard?: boolean;
  filter?: string;
};

export interface CardsState {
  cards: Card[];
  selectedCardId: string | null;
  loading: boolean;
  error: null | string;
}

export type CreateCardValues = {
  title: string;
  type: CardType;
  sequenceOrder: number;
  learningItemId: string | undefined;
  json: CardJson;
  confidenceCheck?: boolean;
};

export type ImportCardResponse = {
  cards: Card[];
};

export type Card = CreateCardValues & {
  id: string;
  learningItem: string;
};

//Content Blocks

export type CardJson = {
  description: string;
  contentBlocks: ContentBlockType[];
  version: string;
  titleSynced?: boolean;
  templateType: TemplateType;
};

export type SlateJSON = Descendant[];

type ContentBlockBase = {
  id: string;
  type: string;
};

export type KnowledgeCheckBlockBase = ContentBlockBase & {
  correctFeedback: Feedback;
  incorrectFeedback: Feedback;
  question: SlateJSON;
};

type TextContentBlockPartial =
  | {
    text: string;
  }
  | {
    json: SlateJSON;
  };

export type TitleBlockType = ContentBlockBase & {
  type: 'title';
} & TextContentBlockPartial;

export type SubtitleBlockType = ContentBlockBase & {
  type: 'subtitle';
} & TextContentBlockPartial;

export type BodyBlockType = ContentBlockBase & {
  type: 'body';
} & TextContentBlockPartial;

export type MediaBlockType = ContentBlockBase & {
  url: string | undefined;
  name: string | undefined;
};

export type ImageBlockType = MediaBlockType & {
  type: 'image';
};

export type VideoBlockType = MediaBlockType & {
  type: 'video';
};

export type AudioBlockType = MediaBlockType & {
  type: 'audio';
};

export type MultipleChoiceBlockType = KnowledgeCheckBlockBase & {
  type: 'multipleChoice';
  multipleChoiceType: MultipleChoiceType;
  options: Array<AnswerOption>;
  randomize: boolean;
};

export type Feedback = {
  header: SlateJSON;
  body: SlateJSON;
};

export type MultipleChoiceType = 'selectone' | 'selectall';

export type TrueFalseBlockType = KnowledgeCheckBlockBase & {
  type: 'trueFalse';
  multipleChoiceType: 'selectone';
  options: [AnswerOption, AnswerOption];
};

export type AnswerOption = {
  id: string;
  optionText: SlateJSON;
  isCorrect: boolean;
};

export type ExpandableListBlockType = ContentBlockBase & {
  type: 'expandableList';
  sections: Array<ListSection>;
};

export type ListSection = {
  id: string;
  title: SlateJSON;
  content: SlateJSON;
};

export type ListSectionType = 'title' | 'content';
export type MultipleChoiceSectionType = 'question' | 'option';

// Types that can be treated as text. IE ones that have text or json properties (per TextContentBlockPartial)
export type TextContentBlockType =
  | TitleBlockType
  | SubtitleBlockType
  | BodyBlockType
  | ExpandableListBlockType
  | MultipleChoiceBlockType
  | TrueFalseBlockType;

export type KnowledgeCheckBlockType =
  | MultipleChoiceBlockType
  | TrueFalseBlockType;

export type TextContentBlockTypes = TextContentBlockType['type'];

export type ContentBlockType =
  | TitleBlockType
  | SubtitleBlockType
  | BodyBlockType
  | ImageBlockType
  | ExpandableListBlockType
  | VideoBlockType
  | AudioBlockType
  | MultipleChoiceBlockType
  | TrueFalseBlockType;

export type ContentBlockTypes = ContentBlockType['type'];

export type TemplateType =
  | 'blank'
  | 'text'
  | 'media'
  | 'expandableList'
  | 'slideshow'
  | 'multipleChoice'
  | 'trueFalse'
  | null;
