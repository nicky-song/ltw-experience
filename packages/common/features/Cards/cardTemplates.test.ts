import {
  getAudioBlock,
  getContentBlockByTemplateType,
  getSubtitleBlock,
  getVideoBlock,
} from './cardTemplates';
import { TemplateType } from './cardTypes';

// Mock uuidv4 so tests are deterministic
jest.mock('uuid', () => ({
  v4: () => 'mocked-id',
}));

describe('Card Templates', () => {
  it('should return the correct content block for all types', () => {
    const contentBlockTypes: TemplateType[] = [
      'blank',
      'text',
      'media',
      'slideshow',
      'expandableList',
    ];
    contentBlockTypes.forEach((type) => {
      expect(getContentBlockByTemplateType(type)).toMatchSnapshot(type);
    });

    expect(getSubtitleBlock('test', 'Subtitle Name')).toMatchSnapshot(
      'subtitle',
    );
    expect(getVideoBlock('test', 'video://url', 'Video Name')).toMatchSnapshot(
      'video',
    );
    expect(getAudioBlock('test', 'audio://url', 'Audio Name')).toMatchSnapshot(
      'audio',
    );
  });
});
