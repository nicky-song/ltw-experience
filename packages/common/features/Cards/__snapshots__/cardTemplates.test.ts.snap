// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Templates should return the correct content block for all types: audio 1`] = `
{
  "id": "test",
  "name": "Audio Name",
  "type": "audio",
  "url": "audio://url",
}
`;

exports[`Card Templates should return the correct content block for all types: blank 1`] = `[]`;

exports[`Card Templates should return the correct content block for all types: expandableList 1`] = `
[
  {
    "id": "mocked-id",
    "json": [
      {
        "children": [
          {
            "text": "Untitled",
          },
        ],
        "type": "title",
      },
    ],
    "type": "title",
  },
  {
    "id": "mocked-id",
    "sections": [
      {
        "content": [
          {
            "children": [
              {
                "text": "Let's enhance the way users comprehend the information by infusing it with your own unique thoughts and ideas. So, go ahead and add your perspective to the mix – it's the key to making our content relatable and engaging!",
              },
            ],
            "type": "paragraph",
          },
        ],
        "id": "mocked-id",
        "title": [
          {
            "children": [
              {
                "text": "Section Header 1",
              },
            ],
            "type": "list-item",
          },
        ],
      },
      {
        "content": [
          {
            "children": [
              {
                "text": "Let's enhance the way users comprehend the information by infusing it with your own unique thoughts and ideas. So, go ahead and add your perspective to the mix – it's the key to making our content relatable and engaging!",
              },
            ],
            "type": "paragraph",
          },
        ],
        "id": "mocked-id",
        "title": [
          {
            "children": [
              {
                "text": "Section Header 2",
              },
            ],
            "type": "list-item",
          },
        ],
      },
      {
        "content": [
          {
            "children": [
              {
                "text": "Let's enhance the way users comprehend the information by infusing it with your own unique thoughts and ideas. So, go ahead and add your perspective to the mix – it's the key to making our content relatable and engaging!",
              },
            ],
            "type": "paragraph",
          },
        ],
        "id": "mocked-id",
        "title": [
          {
            "children": [
              {
                "text": "Section Header 3",
              },
            ],
            "type": "list-item",
          },
        ],
      },
    ],
    "type": "expandableList",
  },
]
`;

exports[`Card Templates should return the correct content block for all types: media 1`] = `
[
  {
    "id": "mocked-id",
    "json": [
      {
        "children": [
          {
            "text": "Untitled",
          },
        ],
        "type": "title",
      },
    ],
    "type": "title",
  },
  {
    "id": "mocked-id",
    "name": "",
    "type": "image",
    "url": "",
  },
  {
    "id": "mocked-id",
    "json": [
      {
        "children": [
          {
            "text": "This is your body text. Ready to bring life to this empty space? Start adding info for your learners here.",
          },
        ],
        "type": "paragraph",
      },
    ],
    "type": "body",
  },
]
`;

exports[`Card Templates should return the correct content block for all types: slideshow 1`] = `[]`;

exports[`Card Templates should return the correct content block for all types: subtitle 1`] = `
{
  "id": "test",
  "json": [
    {
      "children": [
        {
          "text": "Subtitle Name",
        },
      ],
      "type": "subtitle",
    },
  ],
  "type": "subtitle",
}
`;

exports[`Card Templates should return the correct content block for all types: text 1`] = `
[
  {
    "id": "mocked-id",
    "json": [
      {
        "children": [
          {
            "text": "Untitled",
          },
        ],
        "type": "title",
      },
    ],
    "type": "title",
  },
  {
    "id": "mocked-id",
    "json": [
      {
        "children": [
          {
            "text": "This is your body text. Ready to bring life to this empty space? Start adding info for your learners here.",
          },
        ],
        "type": "paragraph",
      },
    ],
    "type": "body",
  },
]
`;

exports[`Card Templates should return the correct content block for all types: video 1`] = `
{
  "id": "test",
  "name": "Video Name",
  "type": "video",
  "url": "video://url",
}
`;
